File: go/authmiddleware/context.go
package authmiddleware

import "context"

type Role string

const (
	RoleCustomer  Role = "customer"
	RoleBarber    Role = "barber"
	RoleShopOwner Role = "shop_owner"
	RoleAdmin     Role = "admin"
)

type User struct {
	ID     string
	Role   Role
	ShopID string
	JTI    string
}

type ctxKey struct{}

func WithUser(ctx context.Context, u User) context.Context {
	return context.WithValue(ctx, ctxKey{}, u)
}

func FromContext(ctx context.Context) (User, bool) {
	u, ok := ctx.Value(ctxKey{}).(User)
	return u, ok
}

File: rust/auth-middleware/src/context.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    Customer,
    Barber,
    ShopOwner,
    Admin,
}

#[derive(Debug, Clone)]
pub struct User {
    pub id: String,
    pub role: Role,
    pub shop_id: Option<String>,
    pub jti: Option<String>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct Claims {
    pub sub: String,
    pub role: Role,
    #[serde(default)]
    pub shop_id: Option<String>,
    #[serde(default)]
    pub jti: Option<String>,
    pub exp: usize,
    pub iat: usize,
    pub iss: String,
    pub aud: String,
}
